census2020/census2020.ipynb

```diff
@@ -0,0 +1,609 @@
+    "\n",
+    "### Categorical Data (Race)\n",
+    "\n",
+    "To accurately visualize the racial demographics, we will use the same color keys for each unique race that was applied to the 2010 dataset. We will also introduce a few new colors to represent the extra racial categories included in this dataset:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(list(df['race'].unique().compute()))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "color_key = {\n",
+    "    'Asian': 'red',\n",
+    "    'Black' : 'lime',\n",
+    "    'Hawaiian Pacific Islander': 'orange',\n",
+    "    'Hispanic or Other' : 'fuchsia',\n",
+    "    'Mixed' : 'purple',\n",
+    "    'Native American' : 'yellow',\n",
+    "    'White' : 'aqua',\n",
+    "    }"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now, we will aggregate the data using the `ds.by()` parameter, which allows us to group the dataset by the `race` column. This will produce separate aggregates for each racial category which will now be colored using the specified `color_key`.\n",
+    "\n",
+    "Additionally, we will enable the **`dynspread=True`** parameter to dynamically adjust the size of rendered points. This ensures that areas with sparse data points remain visible, even when zooming out or when the data points are too densely packed. This dynamic spreading of points enhances visibility without overwhelming the plot, allowing us to capture both the subtle variations in rural areas and the dense clusters in urban regions. \n",
+    "\n",
+    "You can see how this works by zooming into specific regions of the map if you are working in a live Python environment. The plot will dynamically re-render, adjusting the point sizes and density to provide clearer visibility in the zoomed areas."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import holoviews as hv\n",
+    "import xyzservices.providers as xyz\n",
+    "from IPython.display import display\n",
+    "\n",
+    "label = hv.Tiles(xyz.CartoDB.PositronOnlyLabels())\n",
+    "\n",
+    "def plot_city(longitude_range, latitude_range, w=plot_width, h=plot_height, bgcolor='black'):\n",
+    "    \"\"\"Plots a map of the pop density aggregated by race within specified longitude and latitude ranges\"\"\"\n",
+    "    return df.hvplot.points('x', 'y', rasterize=True, dynspread=True, cnorm='eq_hist',\n",
+    "                            xlim=longitude_range, ylim=latitude_range,\n",
+    "                            cmap=color_key, bgcolor=bgcolor, colorbar=False, hover=False,\n",
```

Comment:
I assume both `hover` and `colorbar` should be `True` here, because indeed, without at least one of those, I can't think of a reason to use `rasterize`.

Reply:
Yes true I forgot about the colorbar.

Though I sort of understand why Isaiah would have disabled them, as they're both hard to parse.

![image](https://github.com/user-attachments/assets/3cbe8b0e-036b-478e-ae26-fb9b18a4419a)

Reply:
> Though I sort of understand why Isaiah would have disabled them, as they're both hard to parse.

Yes, that was exactly why.

Reply:
- The hover information is useful, but I agree it should be formatted better. I think @mattpap indicated to me long ago that this approach was a stopgap, if I'm remembering correctly, and that we might be able to format that as a set of key value pairs like x and y are instead. I think that's an issue to file separately. 
- The colorbar is not just hard to parse, but incorrect -- the brightness level indicates _increasing_ population, but the colorbar suggest that it indicates _decreasing_ population. That's either a bug in this notebook or another issue to file. 
- It would be great if we could easily select SI units for the tick formatting, so that 5.000e+5 would show up as 50K, but I don't know if anyone's implemented something like that.

Reply:
> * The colorbar is not just hard to parse, but incorrect -- the brightness level indicates _increasing_ population, but the colorbar suggest that it indicates _decreasing_ population. That's either a bug in this notebook or another issue to file.

The colorbar only becomes _incorrect_ when I add the color_key options to the plot. The previous plots using uniform colors showed more accurate, even if still hard to parse, colorbars. I don't know if it's a bug or what but it was definitely inaccurate/confusing which informed my decision to remove it entirely.

> * It would be great if we could easily select SI units for the tick formatting, so that 5.000e+5 would show up as 500K, but I don't know if anyone's implemented something like that.

I think this is possible in bokeh using a [NumeralTickFormatter](https://docs.bokeh.org/en/latest/docs/reference/models/formatters.html#bokeh.models.NumeralTickFormatter) but I don't know if it is simple to do in hvPlot. I could go some levels lower to implement that in these plots but I fell it's just introducing unnecessary complexity to the code.

Reply:
- For the colorbar colors, there's not a question of degree of accuracy; the colorbar is simply completely wrong. It is not actually reporting how the data is mapping onto brightness or value; it's reversed. Unless there's some obvious mistake in the example code, please file a bug report rather than working around the issue.
- The colorbar ticks are not very useful right now, so if it is possible to provide a simple option like `.opts(colorbar_tick_format=‘0 a’)` and get ticks like '500k', it would be great to do so.  If we have to jump through tons of hoops to get there, like defining custom plot hooks, then I agree with you that doing so is too complex, and then I agree with Maxime that we should disable the colorbar as not being very useful.

Reply:
> The hover information is useful, but I agree it should be formatted better. I think @mattpap indicated to me long ago that this approach was a stopgap, if I'm remembering correctly, and that we might be able to format that as a set of key value pairs like x and y are instead. I think that's an issue to file separately.

I recall there was a discussion about this probably in a meeting, but don't think any conclusions were ever recorded (in an issue or so). It would be best to start such issue/discussion with an example dataset (shape specifically) and expected output shape.

Reply:
@mattpap , I agree a separate issue would be helpful here. What I'd personally like to see in such an issue is that instead of what the hover shows in the screenshot above, it shows:

```
lat: 30.2460° N
lon: 87.7008° W
Black: 157
Hispanic/Other: 3
Mixed: 12
Native American: 2
White: 91
```
(where I've omitted anything with zero count, since the more you zoom in the more sparsely the categories are populated and the less likely that most categories will need to be listed.)

Can you maybe take a look at this PR's code and set up a similar example in pure Bokeh? This one depends on the entire HoloViz stack, and is pretty unwieldy.